hashes/src/lib.rs

```diff
@@ -220,6 +220,16 @@ pub trait GeneralHash: Hash {
         }
         Self::from_engine(engine)
     }
+
+    /// Hashes the entire contents of the `reader`.
+    #[cfg(all(feature = "std", feature = "bitcoin-io"))]
+    fn hash_reader<R: std::io::Read>(reader: &mut R) -> Result<Self, std::io::Error>
+    where <Self as GeneralHash>::Engine: std::io::Write
```

Comment:
Oh, wait, this impl introduces an annoying additional bound on `Self`. Which is not wrong but I wonder if we want to leak the implementation detail. OTOH it's probably better optimized than not having it.

I suspect the most optimal solution would take `BufRead` and just inputted the output of `fill_buf` consuming it entirely.

Reply:
Definitely agreed that it'd be better to drop the `Write` bound on the engine.

I'm skeptical that requiring `BufRead` would be more efficient in-practice than just using `Read` and calling `read()` on a 64-byte (or whatever) array repeatedly, assuming the user was using a buffered reader in the first place. And requiring `BufRead` feels pretty onerous. Though I know we already did it for the consensus decoding traits. Though we also got complaints about that and I'd kinda like to revert it. It's annoying because it means that whenever you have custom Read types you have to implement the `BufRead` trait for it even though none of the `BufRead` methods are even necessary for hashing.

If a user is using non-buffered readers they will have catastrophic performance problems throughout their codebase. I don't really think it's the job of a hashing library to try to prevent that.

Reply:
Note that the real difference comes from not copying the data several times. We wouldn't be using `read` but `fill_buf` and `consume` to avoid copying. The `io::copy` function has to make another buffer and copy the already-copied data before passing it to the hasher. We can avoid this. Note also that `BufRead` trait also bypasses the problem with the current `Read` API that forces people to zero-out the buffers which is wasteful and the only other solution requires nightly.

> Though we also got complaints about that and I'd kinda like to revert it.

I haven't seen any and I'd want to see if they are relevant. Most of the sensible types already implement the trait. Note that this also significantly helps with `push_decode` implementation since then we don't need to keep track of buffering and the whole implementation becomes more performant without using unsound `unsafe` or unstable nightly APIs.

> I don't really think it's the job of a hashing library to try to prevent that.

It's just a side effect of using a more efficient interface. IME it's easy to forget about buffering.

Reply:
Ok, if there are real efficiency improvements then let's do it. ACK using `BufRead`.